File: app/src/main/java/com/example/komuroshun/androidtabui/card/CardListFragment.java

package com.example.komuroshun.androidtabui.card;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.ListFragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.TranslateAnimation;
import android.widget.AdapterView;
import android.widget.ListView;

import com.example.komuroshun.androidtabui.R;

import java.util.List;

/**
 * A simple {@link Fragment} subclass.
 * Activities that contain this fragment must implement the
 * {@link CardListFragment.OnFragmentInteractionListener} interface
 * to handle interaction events.
 * Use the {@link CardListFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class CardListFragment extends ListFragment {
    // TODO: Rename parameter arguments, choose names that match
    // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
//    private static final String ARG_PARAM1 = "param1";
//    private static final String ARG_PARAM2 = "param2";

    // TODO: Rename and change types of parameters
//    private String mParam1;
//    private String mParam2;

    private OnFragmentInteractionListener mListener;

    public CardListFragment() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     //     * @param param1 Parameter 1.
     //     * @param param2 Parameter 2.
     * @return A new instance of fragment CardListFragment.
     */
    // TODO: Rename and change types and number of parameters
    public static CardListFragment newInstance(/*String param1, String param2*/) {
        CardListFragment fragment = new CardListFragment();
        Bundle args = new Bundle();
//        args.putString(ARG_PARAM1, param1);
//        args.putString(ARG_PARAM2, param2);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
//            mParam1 = getArguments().getString(ARG_PARAM1);
//            mParam2 = getArguments().getString(ARG_PARAM2);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_card_list, container, false);
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        // インストールされているアプリを取得
        PackageManager packageManager = getActivity().getPackageManager();
        List<PackageInfo> packageInfoList = packageManager
                .getInstalledPackages(PackageManager.GET_ACTIVITIES);

        CardListAdapter adapter = new CardListAdapter(getActivity());

        // アダプターにアプリ情報を追加
        if (packageInfoList != null) {
            for (PackageInfo info : packageInfoList) {
                adapter.add(info);
            }
        }
        setListAdapter(adapter);

        // 8dp
        int padding = (int) (getResources().getDisplayMetrics().density * 8);
        ListView listView = getListView();
        listView.setPadding(padding, 0, padding, 0);
        listView.setScrollBarStyle(ListView.SCROLLBARS_OUTSIDE_OVERLAY);
        listView.setDivider(null);

        // 余白用のヘッダー、フッターを追加
        LayoutInflater layoutInflater = LayoutInflater.from(getActivity());
        View header = layoutInflater.inflate(R.layout.parts_card_list_header_footer, listView, false);
        View footer = layoutInflater.inflate(R.layout.parts_card_list_header_footer, listView, false);
        listView.addHeaderView(header, null, false);
        listView.addFooterView(footer, null, false);

        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> arg0, View card, int arg2, long arg3) {
                startCardAnimation(card);
            }
        });
    }

    // TODO: Rename method, update argument and hook method into UI event
    public void onButtonPressed(Uri uri) {
        if (mListener != null) {
            mListener.onFragmentInteraction(uri);
        }
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
//        if (context instanceof OnFragmentInteractionListener) {
//            mListener = (OnFragmentInteractionListener) context;
//        } else {
//            throw new RuntimeException(context.toString()
//                    + " must implement OnFragmentInteractionListener");
//        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    private void startCardAnimation (final View card) {
        ListView listView = getListView();
        int[] listViewLocation = new int[2];
        listView.getLocationOnScreen(listViewLocation);
        // get Card coordinate after moving
        int toCardY = listViewLocation[1] + (listView.getHeight()/2)-(card.getHeight()/2);
        // get Card coordinate before moving
        int[] cardLocation = new int[2];
        card.getLocationOnScreen(cardLocation);
        // toCardY - fromCardY
        final int movingAmount = toCardY - cardLocation[1];
        // Setting Animation
        Animation cardMovingAnimation = new TranslateAnimation(Animation.ABSOLUTE, (float)0,
                Animation.ABSOLUTE, (float)0,
                Animation.ABSOLUTE, 0,
                Animation.ABSOLUTE, (float)movingAmount);
        cardMovingAnimation.setDuration(400);
        cardMovingAnimation.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) { }
            @Override
            public void onAnimationEnd(Animation animation) {
                card.setTranslationY(movingAmount);
                flipCard(card);
            }
            @Override
            public void onAnimationRepeat(Animation animation) { }
        });
        card.startAnimation(cardMovingAnimation);
    }

    private void flipCard(View rotationView)
    {
        FlipAnimation flipAnimation = new FlipAnimation(rotationView, rotationView);

        if (rotationView.getVisibility() == View.GONE)
        {
            flipAnimation.reverse();
        }
        rotationView.startAnimation(flipAnimation);
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     * <p>
     * See the Android Training lesson <a href=
     * "http://developer.android.com/training/basics/fragments/communicating.html"
     * >Communicating with Other Fragments</a> for more information.
     */
    public interface OnFragmentInteractionListener {
        // TODO: Update argument type and name
        void onFragmentInteraction(Uri uri);
    }
}

File: app/src/main/java/com/example/komuroshun/androidtabui/card/CardListFragment.kt
package com.example.komuroshun.androidtabui.card

import android.content.Context
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.net.Uri
import android.os.AsyncTask
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.ListFragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AlphaAnimation
import android.view.animation.Animation
import android.view.animation.TranslateAnimation
import android.widget.AdapterView
import android.widget.Button
import android.widget.FrameLayout
import android.widget.ListView

import com.example.komuroshun.androidtabui.R

/**
 * This fragment manage CardListViewModel.
 */
class CardListFragment : ListFragment() {

    //TODO: Dagger2 を使う
    private var mCardListViewModel: CardListViewModel? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater!!.inflate(R.layout.fragment_card_list, container, false)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        mCardListViewModel = CardListViewModel.create(activity!!)

        listAdapter = mCardListViewModel!!.getCardListAdapter()

        // 8dp
        val padding = (resources.displayMetrics.density * 8).toInt()
        listView.setPadding(padding, 0, padding, 0)
        listView.scrollBarStyle = ListView.SCROLLBARS_OUTSIDE_OVERLAY
        listView.divider = null

        // 余白用のヘッダー、フッターを追加
        val layoutInflater = LayoutInflater.from(activity)
        val header = layoutInflater.inflate(R.layout.parts_card_list_header, listView, false)
        //TODO: viewModel で管理する
        //TODO: Visible, Gone を切り替える
        val footer = layoutInflater.inflate(R.layout.parts_card_list_footer, listView, false)
        listView.addHeaderView(header, null, false)
        listView.addFooterView(footer, null, false)
        listView.onItemClickListener = AdapterView.OnItemClickListener { arg0, card, arg2, arg3 -> startCardAnimation(card) }

        val campaignButton = activity!!.findViewById<Button>(R.id.btn_campaign)
        val feedinButton = AlphaAnimation(0f, 1f)
        feedinButton.duration = 100
        feedinButton.setAnimationListener(object : Animation.AnimationListener {
            override fun onAnimationStart(animation: Animation) {}
            override fun onAnimationEnd(animation: Animation) {
                campaignButton.visibility = View.VISIBLE
                // listView.setVisibility(View.GONE);
            }

            override fun onAnimationRepeat(animation: Animation) {}
        })
        campaignButton.startAnimation(feedinButton)
    }

    private fun startCardAnimation(card: View) {
        val listView = listView
        val listViewLocation = IntArray(2)
        listView.getLocationOnScreen(listViewLocation)
        // get Card coordinate after moving
        val toCardY = listViewLocation[1] + listView.height / 2 - card.height / 2
        // get Card coordinate before moving
        val cardLocation = IntArray(2)
        card.getLocationOnScreen(cardLocation)
        // toCardY - fromCardY
        val movingAmount = toCardY - cardLocation[1]
        // Setting Animation
        val cardMovingAnimation = TranslateAnimation(Animation.ABSOLUTE, 0.toFloat(),
                Animation.ABSOLUTE, 0.toFloat(),
                Animation.ABSOLUTE, 0f,
                Animation.ABSOLUTE, movingAmount.toFloat())
        cardMovingAnimation.duration = 400
        cardMovingAnimation.setAnimationListener(object : Animation.AnimationListener {
            override fun onAnimationStart(animation: Animation) {}
            override fun onAnimationEnd(animation: Animation) {
                card.translationY = movingAmount.toFloat()
                flipCard(card)
            }

            override fun onAnimationRepeat(animation: Animation) {}
        })
        card.startAnimation(cardMovingAnimation)
    }

    private fun flipCard(rotationView: View) {
        val flipAnimation = FlipAnimation(rotationView, rotationView)

        if (rotationView.visibility == View.GONE) {
            flipAnimation.reverse()
        }
        rotationView.startAnimation(flipAnimation)
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     */
    interface OnFragmentInteractionListener {
        // TODO: 画面遷移の方法を変える Navigator が使えるか検討
        fun onFragmentInteraction(uri: Uri)
    }

    companion object {
        /**
         * @return A new instance of fragment CardListFragment.
         */
        fun newInstance(): CardListFragment {
            return CardListFragment()
        }
    }
}
